### PR TITLE
SignalR Negotiate protocol versioning

### PR DESCRIPTION
**Pull request by mikaelm12**
This includes the changes for the server and the .NET client. Still polishing but ready for some review
Spec: https://github.com/aspnet/specs/pull/325

### CODE REVIEW

File: src/SignalR/clients/csharp/Client/test/FunctionalTests/Hubs.cs [hunk 1]
```diff
@@ -279,4 +279,14 @@ public class HubWithAuthorization2 : Hub
     {
         public string Echo(string message) => TestHubMethodsImpl.Echo(message);
     }
+
+    public class NegotiateVersion12Hub : Hub
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client/test/FunctionalTests/Hubs.cs]: Don't need these hubs really

**Review by BrennanConroy — COMMENTED** [review]
Quick first pass

File: src/SignalR/clients/csharp/Client/test/FunctionalTests/Hubs.cs (hunk 1 above)
Comment (mikaelm12) [reply src/SignalR/clients/csharp/Client/test/FunctionalTests/Hubs.cs]: Yeah, I guess we can just reuse other ones and ax these

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 2]
```diff
@@ -308,9 +308,23 @@ private async Task ProcessNegotiate(HttpContext context, HttpConnectionDispatche
...
         private static void WriteNegotiatePayload(IBufferWriter<byte> writer, string connectionId, HttpContext context, HttpConnectionDispatcherOptions options)
         {
-            var response = new NegotiationResponse();
-            response.ConnectionId = connectionId;
-            response.AvailableTransports = new List<AvailableTransport>();
+            var response = new NegotiationResponse
+            {
+                ConnectionId = connectionId,
+                AvailableTransports = new List<AvailableTransport>(),
+                Version = options.MinimumProtocolVersion
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs]: Dangling commas? 🤦‍♂

**Comment by halter73** [comment]
It might be nice to rename the "version" query string parameter and negotiate property to "protocolVersion" or similar so customers don't think this is supposed to match the SiganlR package version.

**Review by halter73 — APPROVED** [review]

**Comment by mikaelm12** [comment]
> It might be nice to rename the "version" query string parameter and negotiate property to "protocolVersion"

ProtocolVersion is taken 
https://github.com/aspnet/AspNetCore/blob/bcdc9b08dfbb79ca6b8780e9ef5f9cabba67f71d/src/SignalR/common/Http.Connections.Common/src/NegotiateProtocol.cs#L160-L163

We can use `NegotiateVersion` if no one hates it too much

**Comment by mikaelm12** [comment]
Note to self to add a test for client sending an invalid version in the query string

**Comment by halter73** [comment]
I think `NegotiateVersion` is fine.

**Review by BrennanConroy — COMMENTED** [review]
We need to update the transport spec document too!

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.Log.cs [hunk 3]
```diff
@@ -52,6 +52,12 @@ internal static class Log
             private static readonly Action<ILogger, string, Exception> _failedToReadHttpRequestBody =
                 LoggerMessage.Define<string>(LogLevel.Debug, new EventId(14, "FailedToReadHttpRequestBody"), "Connection {TransportConnectionId} failed to read the HTTP request body.");
 
+            private static readonly Action<ILogger, int, Exception> _negotiateProtocolVersionMismatch =
+                LoggerMessage.Define<int>(LogLevel.Debug, new EventId(14, "NegotiateProtocolVersionMismatch"), "The client requested version '{clientProtocolVersion}', but the server does not support this version.");
```
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.Log.cs]: Change event ids

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 4]
```diff
@@ -306,9 +307,48 @@ private async Task ProcessNegotiate(HttpContext context, HttpConnectionDispatche
...
+                        response.Version = _protocolVersion;
+                    }
+                    else
+                    {
+                        response.Version = clientProtocolVersion;
+                    }
+                }
+                else
+                {
+                    response.Error = $"The client requested an invalid protocol version '{queryStringVersionValue}'";
```
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs]: Should have a test for this case

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 5]
```diff
@@ -306,9 +307,48 @@ private async Task ProcessNegotiate(HttpContext context, HttpConnectionDispatche
...
+                    Log.InvalidNegotiateProtocolVersion(_logger, queryStringVersionValue);
+                    NegotiateProtocol.WriteResponse(response, writer);
+                    return;
+                }
+            }
+            else if (options.MinimumProtocolVersion > 0)
+            {
+                // NegotiateVersion wasn't parsed meaning the client requests version 0.
+                response.Error = $"The client requested version '0', but the server does not support this version.";
```
Comment (BrennanConroy) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs]: And a test for this

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs]: Preferably one of these tests would verify there is no connection ID in the response payload

**Review by BrennanConroy — APPROVED** [review]

**Review by BrennanConroy — APPROVED** [review]

**Comment by mikaelm12** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=336108&view=logs

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3057

**Comment by mikaelm12** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=336109&view=logs

**Pull request merged**